crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1220 @@
+        object_id: &ObjectID,
+        version: SequenceNumber,
+    ) -> SuiResult<Option<ObjectRef>> {
+        // There is probably a more efficient way to implement this, but since this is only used by
+        // old protocol versions, it is better to do the simple thing that is obviously correct.
+        // In this case we previous version from all sources and choose the highest
+        let mut candidates = Vec::new();
+
+        let check_versions =
+            |versions: &CachedVersionMap<ObjectEntry>| match versions.get_prior_to(&version) {
+                Some((version, object_entry)) => match object_entry {
+                    ObjectEntry::Object(object) => {
+                        assert_eq!(object.version(), version);
+                        Some(object.compute_object_reference())
+                    }
+                    ObjectEntry::Deleted => {
+                        Some((*object_id, version, ObjectDigest::OBJECT_DIGEST_DELETED))
+                    }
+                    ObjectEntry::Wrapped => {
+                        Some((*object_id, version, ObjectDigest::OBJECT_DIGEST_WRAPPED))
+                    }
+                },
+                None => None,
+            };
+
+        // first check dirty data
+        if let Some(objects) = self.dirty.objects.get(object_id) {
+            if let Some(prior) = check_versions(&objects) {
+                candidates.push(prior);
+            }
+        }
+
+        if let Some(objects) = self.cached.object_cache.get(object_id) {
+            if let Some(prior) = check_versions(&objects.lock()) {
+                candidates.push(prior);
+            }
+        }
+
+        if let Some(prior) = self
+            .store
+            .get_object_ref_prior_to_key_deprecated(object_id, version)?
+        {
+            candidates.push(prior);
+        }
+
+        // sort candidates by version, and return the highest
+        candidates.sort_by_key(|(_, version, _)| *version);
+        Ok(candidates.pop())
+    }
+
+    fn get_root_state_accumulator_for_epoch(
+        &self,
+        epoch: EpochId,
+    ) -> SuiResult<Option<(CheckpointSequenceNumber, Accumulator)>> {
+        self.store.get_root_state_accumulator_for_epoch(epoch)
+    }
+
+    fn get_root_state_accumulator_for_highest_epoch(
+        &self,
+    ) -> SuiResult<Option<(EpochId, (CheckpointSequenceNumber, Accumulator))>> {
+        self.store.get_root_state_accumulator_for_highest_epoch()
+    }
+
+    fn insert_state_accumulator_for_epoch(
+        &self,
+        epoch: EpochId,
+        checkpoint_seq_num: &CheckpointSequenceNumber,
+        acc: &Accumulator,
+    ) -> SuiResult {
+        self.store
+            .insert_state_accumulator_for_epoch(epoch, checkpoint_seq_num, acc)
+    }
+
+    fn iter_live_object_set(
+        &self,
+        include_wrapped_tombstone: bool,
+    ) -> Box<dyn Iterator<Item = LiveObject> + '_> {
+        self.store.iter_live_object_set(include_wrapped_tombstone)
```

Comment:
Should this also read from the cache? If not, consider adding a comment about why not.

Reply:
It should not - the only time it is safe to iterate the live object set is at an epoch boundary, at which point the db is consistent and the dirty cache is empty.

Reply:
added a comment